kernel/src/table_changes/data_read.rs

```diff
@@ -0,0 +1,226 @@
+use std::collections::HashMap;
+use std::iter::{empty, once};
+use std::sync::Arc;
+
+use itertools::{Either, Itertools};
+use roaring::RoaringTreemap;
+use url::Url;
+
+use crate::actions::deletion_vector::{split_vector, treemap_to_bools};
+use crate::expressions::{column_expr, ColumnName, Scalar};
+use crate::scan::state::DvInfo;
+use crate::scan::{parse_partition_value, ColumnType, ScanResult};
+use crate::schema::{DataType, SchemaRef, StructField, StructType};
+use crate::{DeltaResult, Engine, Error, Expression, FileMeta};
+
+use super::scan::GlobalScanState;
+use super::scan_file::{ScanFile, ScanFileType};
+type ResolvedScanFile = (ScanFile, Option<Vec<bool>>);
+
+pub(crate) struct ScanFileReader {
+    global_scan_state: GlobalScanState,
+    scan_file: ScanFile,
+    selection_vector: Option<Vec<bool>>,
+}
+
+impl ScanFileReader {
+    pub(crate) fn new(
+        global_scan_state: GlobalScanState,
+        scan_file: ScanFile,
+        selection_vector: Option<Vec<bool>>,
+    ) -> Self {
+        Self {
+            global_scan_state,
+            scan_file,
+            selection_vector,
+        }
+    }
+
+    pub(crate) fn get_generated_columns(&self) -> DeltaResult<HashMap<String, Expression>> {
+        // Both in-commit timestamps and file metadata are in milliseconds
+        //
+        // See:
+        // [`FileMeta`]
+        // [In-Commit Timestamps] : https://github.com/delta-io/delta/blob/master/PROTOCOL.md#writer-requirements-for-in-commit-timestampsa
+        let timestamp = Scalar::timestamp_from_millis(self.scan_file.timestamp)?;
+        let commit_version: i64 = self
+            .scan_file
+            .commit_version
+            .try_into()
+            .map_err(Error::generic)?;
+        let cols = ["_change_type", "_commit_version", "_commit_timestamp"];
+        let expressions = match self.scan_file.tpe {
+            ScanFileType::Cdc => [
+                column_expr!("_change_type"),
+                Expression::literal(commit_version),
+                timestamp.into(),
+            ],
+            ScanFileType::Add => [
+                "insert".into(),
+                Expression::literal(commit_version),
+                timestamp.into(),
+            ],
+
+            ScanFileType::Remove => [
+                "delete".into(),
+                Expression::literal(commit_version),
+                timestamp.into(),
+            ],
+        };
+        let generated_columns: HashMap<String, Expression> = cols
+            .iter()
+            .map(ToString::to_string)
+            .zip(expressions)
+            .collect();
+        Ok(generated_columns)
```

Comment:
No need to make these `Strings`:
```suggestion
        Ok(cols.into_iter().zip(expressions).collect())
```

Also requires:
- Change the return of this to `DeltaResult<HashMap<&str, Expression>>`
- do: `.remove(field_name.as_str())` on 106 below